src/seedu/addressbook/commands/TagCommand.java

```diff
@@ -0,0 +1,75 @@
+package seedu.addressbook.commands;
+
+import seedu.addressbook.common.Messages;
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.data.person.UniquePersonList;
+import seedu.addressbook.data.person.UniquePersonList.DuplicatePersonException;
+import seedu.addressbook.data.person.UniquePersonList.PersonNotFoundException;
+import seedu.addressbook.data.tag.Tag;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Add and remove tags on a person using it's last displayed index from the address book.
+ */
+public class TagCommand extends Command {
+
+    public static final String COMMAND_WORD = "tag";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Add and remove tags on the person identified by the index number used in the last person listing.\n"
+            + "Parameters: INDEX +TAG [-TAG ...]\n"
+            + "Example: tag 1 -owesMoney +bestFriend";
+
+    public static final String MESSAGE_TAG_PERSON_SUCCESS = "Updated Tags on Person: %1$s";
+
+    public static final String MESSAGE_TAG_DUPLICATE_PERSON = "Update tags failed: %1$s\n";
+
+
+    private final Set<Tag> tagsToAdd = new HashSet<>();
+    private final Set<Tag> tagsToDelete = new HashSet<>();
+
+    public Set<Tag> getTagsToAdd() {
+        return tagsToAdd;
+    }
+
+    public Set<Tag> getTagsToDelete() {
+        return tagsToDelete;
+    }
+
+    public TagCommand(int targetVisibleIndex, Set<String> tagStringsToAdd, Set<String> tagStringsToDelete) throws IllegalValueException {
+        super(targetVisibleIndex);
+        for (String tagString : tagStringsToAdd) {
+            tagsToAdd.add(new Tag(tagString));
+        }
+        for (String tagString : tagStringsToDelete) {
+            tagsToDelete.add(new Tag(tagString));
+        }
+    }
+
+    @Override
+    public CommandResult execute() {
+        try {
+            final ReadOnlyPerson target = getTargetPerson();
+            Person targetWithUpdatedTags = new Person(target);
+            Set<Tag> tagSet = targetWithUpdatedTags.getTags();
+            tagSet.removeAll(tagsToDelete);
+            tagSet.addAll(tagsToAdd);
+            targetWithUpdatedTags.setTags(tagSet);
+            addressBook.replacePerson(target, targetWithUpdatedTags);
+
+            return new CommandResult(String.format(MESSAGE_TAG_PERSON_SUCCESS, targetWithUpdatedTags));
```

Comment:
When you change the same person's tags in two separate commands, the second command will complain that the person cannot be found. Example:
`tag 1 +asd`
`tag 1 -asd`

This is caused by this line, which causes `Main` to not update the `lastShownList` even though tags have been changed for the target person once.

As a result, the subsequent execution of `executeCommand()` in `Main` sets the outdated ` List<ReadOnlyPerson>` to the `tagCommand`, resulting in `tagCommand` to find the wrong `ReadOnlyPerson` with `getTargetPerson()`.

Reply:
Changing the line to 
`return new CommandResult(String.format(MESSAGE_TAG_PERSON_SUCCESS, targetWithUpdatedTags), addressBook.getAllPersons().immutableListView());`
should be an easy fix, but causes the list to be displayed again, causing I/O tests to fail (JUnit tests still fine).

Reply:
Nice find! Thanks! Indeed a successful `tag` command will invalidate the contents of `Main#lastShownList`. However I don't think we should be returning the list of all persons of the entire address book as it will also change the `lastShownList` context in an unexpected way.
Do you think it makes more sense to also perform the tag update in `Command#relevantPersons` then return that result back up into `Main`?

Reply:
Right that makes sense! I agree, the context of the `lastShownList` could have been that of after a `find` command, in which case only a subset of the addressbook should be shown.

I think your solution works better, I agree with changing `relevantPersons` in the `tagCommand`!